dev/com.ibm.ws.messaging.msgstore/src/com/ibm/ws/sib/msgstore/expiry/Expirer.java

```diff
@@ -91,101 +91,67 @@
 
     /**
      * Constructor
-     * @param ms the MessageStore  
+     * @param messageStore the MessageStore  
      */
-    public Expirer(MessageStoreImpl ms)
-    {
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) SibTr.entry(this, tc, "<init>");
+    public Expirer(MessageStoreImpl messageStore) {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) SibTr.entry(this, tc, "<init>", messageStore);
 
-        messageStore = ms;
-        expiryIndex = new ExpiryIndex();
-
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) SibTr.exit(this, tc, "<init>", this);
+        this.messageStore = messageStore;
+       
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) SibTr.exit(this, tc, "<init>");
     }
 
     /**
-     * Add an Expirable reference for an item to the expiry index. The reference will be 
-     * added to the index in order of expiry time (which must be set within the Expirable).
-     * Once added to the index, it will become eligible for expiry processing at the 
-     * designated time.
-     * 
-     * @param expirable the Expirable item for which a reference is to be added to the expiry index.
+     * Include a WeakReference to an Expirable in the set of pending Expirables. 
      * 
-     * @return true if the reference was added to the index, false otherwise. False may
-     * be returned if the item has not yet been added to an item stream and therefore does not 
-     * have a unique ID. False may also be returned if the expirer has not been started.
-     * @throws SevereMessageStoreException 
+     * @param expirable the Expirable to be added to the set of Expirables.
+     * @throws SevereMessageStoreException if the Expirable is already added.
      */ 
-    public final boolean addExpirable(Expirable expirable) throws SevereMessageStoreException
+    public final void addExpirable(Expirable expirable) throws SevereMessageStoreException
     {
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
         {
             SibTr.entry(tc, "addExpirable",
                        "objId="
                        + (expirable == null ? "null" : String.valueOf(expirable.expirableGetID()))
-                       + " addEnabled="
-                       + addEnabled);
+                       + " started="+started);
         }
 
-        boolean reply = false;
-
         // Ignore this entry if the expirer has ended or the given entry is null
-        if (addEnabled && expirable != null)
-        {
-            long expiryTime = expirable.expirableGetExpiryTime();
-
+        // or the referenced item has already gone from the message store.     
+        if (started.get() && expirable != null && expirable.expirableIsInStore()){
```

Comment:
Since expirable is a parameter and can't be modified by another thread, there is no reason to place the AtomicBoolean.get() first — it forces a JVM memory read barrier but this can't affect the value of the next check in the logical conjunction.

If expirable is likely to be null, perhaps we should test that first, and avoid any additional expense?